src/Npgsql/NpgsqlNestedDataReader.cs

```diff
@@ -0,0 +1,509 @@
+﻿using Npgsql.Internal;
+using Npgsql.Internal.TypeHandlers;
+using Npgsql.Internal.TypeHandling;
+using Npgsql.PostgresTypes;
+using Npgsql.TypeMapping;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Data.Common;
+using System.IO;
+using System.Runtime.CompilerServices;
+
+namespace Npgsql
+{
+    /// <summary>
+    /// Reads a forward-only stream of rows from a nested data source.
+    /// Can be retrieved using <see cref="NpgsqlDataReader.GetData(int)"/> or
+    /// <see cref="NpgsqlNestedDataReader.GetData(int)"/>.
+    /// </summary>
+    public sealed class NpgsqlNestedDataReader : DbDataReader
+    {
+        NpgsqlDataReader _outermostReader;
+        ulong _uniqueOutermostReaderRowId;
+        NpgsqlNestedDataReader? _outerNestedReader;
```

Comment:
This doesn't seem to get assigned anywhere...

Also, we can just have an `object` `_parentReader` which points to either a `NpgsqlNestedDataReader` or `NpgsqlDataReader`, and switch on the type in Dispose.

Reply:
Not really, we need to have a reference to the outermost reader, so that we can fetch the Buffer and the Connector.

Reply:
I think I meant one object _parentReader in addition to _outermostReader, which then you can switch on for type in Dispose - but on second look not very significant.